src/smart.ts

```diff
@@ -134,6 +138,35 @@ function any(tasks: Task[]): Promise<any> {
     });
 }
 
+/**
+ * The maximum length for a code verifier for the best security we can offer.
+ * Please note the NOTE section of RFC 7636 § 4.1 - the length must be >= 43,
+ * but <= 128, **after** base64 url encoding. This means 32 code verifier bytes
+ * encoded will be 43 bytes, or 96 bytes encoded will be 128 bytes. So 96 bytes
+ * is the highest valid value that can be used.
+ */
+ var RECOMMENDED_CODE_VERIFIER_LENGTH = 96;
+
+ /**
+  * Character set to generate code verifier defined in rfc7636.
+  */
+ var PKCE_CHARSET = 'ABCDEFGHIJKLMNOPQRSTUVWXYZabcdefghijklmnopqrstuvwxyz0123456789-._~';
+
+ /**
+  * Generates a code_verifier and code_challenge, as specified in rfc7636.
+  */
+ function generatePKCECodes() {
+   var inputBytes:Buffer = jose.util.randomBytes(RECOMMENDED_CODE_VERIFIER_LENGTH);
+   var input:string = Array.from(inputBytes).map(function(val:number) { return PKCE_CHARSET[val % PKCE_CHARSET.length]; }).join('');
+   var codeVerifier:string = jose.util.base64url.encode(input);
```

Comment:
This is base64urlencoding a string that's already (by  construction) a valid base64urlencoding. You're not getting the full entropy of your random bytes because of the `%` operation. Cleaner to do:

```
codeVerifier: string = jose.util.base64url.encode(jose.util.randomBytes(RECOMMENDED_CODE_VERIFIER_LENGTH))
```